File: backend/app/features/chat/egress_scan.py
"""Heuristic secret-pattern scanner for the cloud egress-friction layer (Slice 14, §5.1).

This is *friction, not redaction* and *not* a DLP product (§5.5): a precision-over-recall
regex pass that FLAGS likely-secret content so the UI can ask "send anyway?" before a
``cloud_enabled`` engagement's message leaves the local network. A missed secret is a known
limitation; a false-positive modal is mere annoyance (Risk 4).

The matched VALUE is never captured, returned, logged, or persisted — only the stable
category NAME — so a secret can never leak into the audit log, a 409 body, or a log line via
this module (§5.5 / Risk 7). ``EgressMatch`` deliberately carries no substring field.

The v1 pattern set is LOCKED to the §5.1 examples (Resolved decision 2): no GitHub PAT, GCP
SA JSON, or Stripe key in this slice. The frontend re-implements a SUBSET of these for a
pre-flight UX scan; THIS module is the single server-authoritative source of truth — the POST
guard re-scans regardless of the client (Risk 3). Patterns are tuned precision-over-recall.
"""

from __future__ import annotations

import re
from dataclasses import dataclass

__all__ = ["EgressMatch", "scan", "category_names"]


@dataclass(frozen=True)
class EgressMatch:
    """One matched secret-pattern category.

    Carries ONLY the stable category name — never the matched substring (§5.5 / Risk 7).
    Keeping the value out of this object keeps it out of every downstream sink (audit
    payload, 409 body, logs), since there is nothing to leak."""

    category: str


# Ordered, named patterns. ``scan`` reports categories in this declaration order so the
# modal copy / audit list is stable. Each is anchored / length-bounded for precision: the
# goal is to almost never fire on ordinary prose, accepting that a cleverly-formatted secret
# may slip past (friction, not DLP — §5.1 / Risk 4).
_PATTERNS: tuple[tuple[str, re.Pattern[str]], ...] = (
    # AWS access key id: AKIA/ASIA prefix + 16 uppercase base32 chars (20 total).
    ("aws_access_key", re.compile(r"\b(?:AKIA|ASIA)[A-Z0-9]{16}\b")),
    # PEM private-key header, any algorithm label (RSA/EC/OPENSSH/ENCRYPTED/…) or none.
    ("private_key_block", re.compile(r"-----BEGIN(?: [A-Z0-9]+)* PRIVATE KEY-----")),
    # JWT: three base64url segments separated by dots; header segment starts ``eyJ`` (the
    # base64url of ``{"``), which is what distinguishes a JWT from any dotted token.
    ("jwt", re.compile(r"\beyJ[A-Za-z0-9_-]+\.[A-Za-z0-9_-]+\.[A-Za-z0-9_-]+")),
    # password=/passwd:/pwd = assignment with a non-empty value. The required [:=] keeps
    # bare prose ("I forgot my password") from matching (precision negative, Risk 4).
    ("password_assignment", re.compile(r"(?i)\b(?:password|passwd|pwd)\s*[:=]\s*\S+")),
    # api_key/apikey/secret/token assignment with a LONG (>=16) opaque value, so "token=1"
    # or "the secret is safe" do not fire (precision).
    (
        "generic_api_key",
        re.compile(r"(?i)\b(?:api[_-]?key|secret|token)\s*[:=]\s*['\"]?[A-Za-z0-9_\-./+]{16,}"),
    ),
    # Authorization: Bearer <token> — require a long opaque token so "bearer of news" misses.
    ("bearer_token", re.compile(r"(?i)\bBearer\s+[A-Za-z0-9_\-.=+/]{12,}")),
    # Slack token: xoxb-/xoxa-/xoxp-/xoxr-/xoxs- + a long body (high-precision representative).
    ("slack_token", re.compile(r"\bxox[baprs]-[A-Za-z0-9-]{10,}")),
)


def scan(content: str) -> list[EgressMatch]:
    """Scan ``content`` for likely-secret patterns; return the matched categories.

    Returns one :class:`EgressMatch` per matched category, in pattern-declaration order,
    deduplicated (a category matched by several substrings is reported once). Returns ``[]``
    when nothing matches. The matched substring is intentionally NOT returned (§5.5)."""
    matches: list[EgressMatch] = []
    for category, pattern in _PATTERNS:
        if pattern.search(content):
            matches.append(EgressMatch(category=category))
    return matches


def category_names(content: str) -> list[str]:
    """Convenience: the matched category names only (e.g. for the 409 body / audit list).

    Pure projection of :func:`scan` — still never the matched value (§5.5)."""
    return [m.category for m in scan(content)]

File: frontend/src/features/chat/egressScan.ts
/**
 * Client-side pre-flight mirror of a SUBSET of the backend egress secret-pattern scanner
 * (Slice 14, §5.1) — see `backend/app/features/chat/egress_scan.py`.
 *
 * This is convenience UX only: it decides whether to show the friction modal *before* the
 * POST, so the user is asked before anything leaves the machine. The SERVER is authoritative —
 * it re-scans every cloud_enabled send and returns a 409 regardless of the client (Risk 3), so
 * a client/server pattern drift can never let an unconfirmed secret reach the cloud.
 *
 * Returns matched category NAMES only — never the matched value (§5.5). The names match the
 * backend exactly so the modal labels line up with the server's 409 `matched_categories`.
 */

interface EgressPattern {
  readonly category: string
  readonly pattern: RegExp
}

// A precision-tuned SUBSET of the backend's locked v1 set (slack_token is server-only). Each
// pattern mirrors its Python counterpart; JS uses the `i` flag where Python used `(?i)`.
const PATTERNS: readonly EgressPattern[] = [
  { category: 'aws_access_key', pattern: /\b(?:AKIA|ASIA)[A-Z0-9]{16}\b/ },
  { category: 'private_key_block', pattern: /-----BEGIN(?: [A-Z0-9]+)* PRIVATE KEY-----/ },
  { category: 'jwt', pattern: /\beyJ[A-Za-z0-9_-]+\.[A-Za-z0-9_-]+\.[A-Za-z0-9_-]+/ },
  { category: 'password_assignment', pattern: /\b(?:password|passwd|pwd)\s*[:=]\s*\S+/i },
  {
    category: 'generic_api_key',
    pattern: /\b(?:api[_-]?key|secret|token)\s*[:=]\s*['"]?[A-Za-z0-9_\-./+]{16,}/i,
  },
  { category: 'bearer_token', pattern: /\bBearer\s+[A-Za-z0-9_\-.=+/]{12,}/i },
]

/**
 * Human-readable labels for the friction modal copy. Covers ALL backend categories (including
 * the server-only `slack_token`) so a modal driven by the server's 409 categories renders
 * properly even for a pattern the client does not itself scan. Unknown categories fall back to
 * the raw name (forward-compatible with a future backend addition).
 */
export const EGRESS_CATEGORY_LABELS: Readonly<Record<string, string>> = {
  aws_access_key: 'AWS access key',
  private_key_block: 'private key block',
  jwt: 'JWT',
  password_assignment: 'password= assignment',
  generic_api_key: 'API key / secret',
  bearer_token: 'bearer token',
  slack_token: 'Slack token',
}

/** Map a category name to its human-readable label (falls back to the raw name). */
export function egressCategoryLabel(category: string): string {
  return EGRESS_CATEGORY_LABELS[category] ?? category
}

/**
 * Scan composer text for likely-secret patterns; return the matched category names
 * (deduplicated, in declaration order). Empty when nothing matches. Never returns the matched
 * value (§5.5) — only the category name.
 */
export function scanEgress(content: string): string[] {
  return PATTERNS.filter(({ pattern }) => pattern.test(content)).map(({ category }) => category)
}
